### PR TITLE
Permit a new array model to have a super set of array model fields.

### PR DESCRIPTION
**Pull request by stuartarchibald**
In the case where a user defines a new array model that extends
from or duck types to Numba's default array model, it should be
valid for the new array model to have additional fields and still
be accepted by Numba's array related implementations.

This patch fixes a check to permit the reuse of the somewhat
fundamental ```numba.np.arrayobj.populate_array``` function.

### CODE REVIEW

File: numba/np/arrayobj.py [hunk 1]
```diff
@@ -221,10 +221,12 @@ def populate_array(array, data, shape, strides, itemsize, meminfo,
...
 
-    # Make sure that we have all the fields
+    # Make sure that we have all the fields needed to wire up a minimal viable
+    # array structure.
     got_fields = set(attrs.keys())
-    if got_fields != required_fields:
-        raise ValueError("missing {0}".format(required_fields - got_fields))
+    for got in got_fields:
+        if got not in required_fields:
+            raise ValueError("missing {0}".format(required_fields - got_fields))
```
[the remark this answers is not in the document]
Comment (sklam) [reply numba/np/arrayobj.py:229]: doesn't the code already do that?

[the remark this answers is not in the document]
Comment (stuartarchibald) [reply numba/np/arrayobj.py:229]: I'm ok with leaving this as-is and revisiting with unit tests at a later time.

**Comment by sklam** [comment]
This diff is all it needs to make the existing subarray tests check the effect of this patch:

```diff
diff --git a/numba/tests/test_ndarray_subclasses.py b/numba/tests/test_ndarray_subclasses.py
index 7644bef9c..07be35776 100644
--- a/numba/tests/test_ndarray_subclasses.py
+++ b/numba/tests/test_ndarray_subclasses.py
@@ -123,7 +123,22 @@ def typeof_ta_ndarray(val, c):
     return MyArrayType(dtype, val.ndim, layout, readonly=readonly)


-register_model(MyArrayType)(numba.core.datamodel.models.ArrayModel)
+@register_model(MyArrayType)
+class TheArrayModel(numba.core.datamodel.models.StructModel):
+    def __init__(self, dmm, fe_type):
+        ndim = fe_type.ndim
+        members = [
+            ('meminfo', types.MemInfoPointer(fe_type.dtype)),
+            ('parent', types.pyobject),
+            ('nitems', types.intp),
+            ('itemsize', types.intp),
+            ('data', types.CPointer(fe_type.dtype)),
+            ('shape', types.UniTuple(types.intp, ndim)),
+            ('strides', types.UniTuple(types.intp, ndim)),
+            ('extra', types.intp),
+        ]
+        super(TheArrayModel, self).__init__(dmm, fe_type, members)
+


 @type_callable(MyArray)
```

**Comment by stuartarchibald** [comment]
> This diff is all it needs to make the existing subarray tests check the effect of this patch:

Thanks, that should exercise it. I've thought about this a bit more and wonder if something like: https://github.com/numba/numba/commit/92c3f563071c66a3be13f094106012a3938fc83b might be a better way to do this. What it does is fix the definition of `required_fields` to find the fields that are needed by the standard array data model in the current context, and then reports on whether the presented array model has fields that match (at least in name). The check here: https://github.com/numba/numba/blob/974a5ca78182d2b35f32df6b395d38ef636f26aa/numba/np/arrayobj.py#L224-L227 is left as-is because it is essentially checking that the code in the function has managed to create a dictionary suitable for wiring through to the standard array model fields, which seems like a different concern. Further, with the fix to the definition of `required_fields`, this check will "just work" conditional on the function being implemented correctly.

**Comment by sklam** [comment]
https://github.com/numba/numba/commit/92c3f563071c66a3be13f094106012a3938fc83b is good. If you commit, we can get it merge quickly.

We can continue to improve it. As I mention out-of-band, we would want to make an easier API for extending the data-model. Perhaps, it also should add something to track how two structs are related (i.e. one is an extension of the other) to make the check here even shorter. Anyway, that's a different PR.

**Pull request merged**